### ISSUE TITLE
Some complex `when/2` conditions are displayed as multiple residual goals

### ISSUE THREAD

**Issue by dgelessus** [SWI-Prolog/swipl-devel#818]
Some complex `when` conditions with conjunctions nested in disjunctions currently result in an extra residual goal:

```prolog
?- when(ground(A) ; ((ground(B), ground(C))), Whened = yes).
when((ground(A);ground(B), ground(C)), Whened=yes),
when((ground(B), ground(C)), when:check_disj(-, [ground(A),  (ground(B), ground(C))], user:(Whened=yes))).
```

The first goal is the expected representation of the when call. The second goal is redundant as it shows only information that the first goal already contains, and it exposes an internal call used by the `when` implementation.

Worse, the second residual goal is displayed even after the `when` condition has already been fulfilled:

```prolog
?- when(ground(A) ; ((ground(B), ground(C))), Whened = yes), A = x.
A = x,
Whened = yes,
when((ground(B), ground(C)), when:check_disj(-, [ground(x),  (ground(B), ground(C))], user:(yes=yes))).
```

This is causing issues with some of our unit tests that check that certain calls succeed without residual goals, to verify that there are no remaining `when`/`freeze`/`block` goals that haven't been executed yet.

---

I had a relatively simple fix for this that I was going to submit (dgelessus/swipl-devel@c8322b98added0d6d37bb1a0a358c2e64e79794f), but after some more testing it turns out that this fix isn't sufficient. I'm not super familiar with SWI's `when` implementation, but it seems that by nesting `when` conditions in the right way, it's possible to build arbitrarily deep chains of `when:trigger` and `when:check_disj` that the current `when_goals` code can't handle.

```prolog
?- when(ground(A) ; ((ground(B), ground(C)), ground(D)), Whened = yes).
when((ground(A);(ground(B), ground(C)), ground(D)), Whened=yes),
when((ground(B), ground(C)), when:trigger(ground(D), when:check_disj(-, [ground(A),  ((ground(B), ground(C)), ground(D))], user:(Whened=yes)))).

?- when(ground(A) ; (((ground(B), ground(C)), ground(D)), ground(E)), Whened = yes).
when((ground(A);((ground(B), ground(C)), ground(D)), ground(E)), Whened=yes),
when((ground(B), ground(C)), when:trigger(ground(D), when:trigger(ground(E), when:check_disj(-, [ground(A),  (((ground(B), ground(C)), ground(D)), ground(E))], user:(Whened=yes))))).
```

A proper fix would probably require making `when_goals` completely recursive.

**Comment by JanWielemaker** [comment]
Good question.   I think you are more familiar with the implementation than I am.   Tom has done the first.   I moved the condition analysis to C and most of the comments and Markus did the attribute_goals//1.  The overall idea on how it works is not that hard to grasp, but the details are tricky.

Note that it seems that disjunctions are designed to leave attributed variables around as residuals.   I'm afraid there is no simple fix.  Well, strictly speaking there is no bug either except that one may run out of resources that could be avoided.   I'm afraid I'm a little too busy to dive into this shortly.   Can we discuss how to deal with this offline?

**Comment by dgelessus** [comment]
Thanks for the insight. To clarify, the issue in my case here is not that `when` leaves attributes around after the goal has executed - as you say, that behavior isn't really incorrect and is just a consequence of how SWI implements `when`.

What *is* an issue is that `when:attribute_goals//1` doesn't handle the extra attributes consistently. For simple disjunctions, the extra attributes are correctly hidden, so that the user sees just a single `when(...)` residual goal that disappears once the condition has fired. The bug I'm describing is that with complex enough conditions, this special handling stops working, and `when:attribute_goals//1` starts returning extra goals in addition to the proper representation (and these extra goals don't disappear after the condition has fired).

So I *think* there are no deep issues with the implementation - probably the existing special cases for handling disjunctions just need to be applied more thoroughly. I'll take another look this week (probably) and will send a PR if I get a better fix working.

**Comment by dgelessus** [comment]
Finally had some time to look into this again properly - sorry for the long silence.

> So I _think_ there are no deep issues with the implementation

Looks like I was wrong - there is a relatively significant issue. In short, if you call `frozen/2` on a variable that appears in a `when` condition disjunction, the `when` effectively disappears and will not execute once the condition becomes true. Example:

```prolog
?- when((nonvar(X) ; nonvar(Y)), format("when goal has run: ~w ~w~n", [X, Y])).
when((nonvar(X);nonvar(Y)), format("when goal has run: ~w ~w~n", [X, Y])).
% OK: condition not true yet, so the when is still there

?- when((nonvar(X) ; nonvar(Y)), format("when goal has run: ~w ~w~n", [X, Y])), X = 1.
when goal has run: 1 _13470
X = 1.
% OK: condition became true, so the when executed and then disappeared

?- when((nonvar(X) ; nonvar(Y)), format("when goal has run: ~w ~w~n", [X, Y])), frozen(X, XGoals).
XGoals = when((nonvar(X);nonvar(Y)), user:format("when goal has run: ~w ~w~n", [X, Y])).
% Bad: condition not true yet, but the when disappeared already!

?- when((nonvar(X) ; nonvar(Y)), format("when goal has run: ~w ~w~n", [X, Y])), frozen(X, XGoals), X = 1.
X = 1,
XGoals = when((nonvar(1);nonvar(Y)), user:format("when goal has run: ~w ~w~n", [1, Y])).
% Bad: condition became true, but the when didn't execute!
```

This only happens when using `frozen/2`. When using `copy_term/3` instead, everything behaves as expected.

As far as I can tell, this has to do with the inconsistent scoping of attribute bindings in `attribute_goals//1` (see #828). To avoid rendering the same `when` call twice, the logic in `when:attribute_goals//1` reuses the shared `DisjVar` variable. Normally that variable is used to track that the same `when` is not executed more than once if multiple parts of its disjunction become true - if the variable is bound, the `when` has already executed once and shouldn't be executed again. But here the same variable is also bound in `attribute_goals` to track that a `when` has been rendered already and shouldn't be rendered a second time. The implementation apparently expects this binding of `DisjVar` inside `attribute_goals` to be temporary, as is the case with `copy_term/3`. But when using `frozen/2`, the `DisjVar` binding instead becomes permanent, which incorrectly marks the `when` as "already executed".

I don't know what would be the best way to fix this, and I think it would depend on the answer to #828. The current `attribute_goals` implementation assumes that it always receives a fresh temporary copy of the attvar. If that behavior can be guaranteed for `copy_term` *and* `frozen`, then `attribute_goals` will work as expected again. But if this is not guaranteed, then `attribute_goals` needs some other way to store temporary state that is scoped to the current `copy_term`/`frozen` call.

Or failing all that, give up on trying to make `attribute_goals` fully reconstruct the original `when` call, and instead return internal goals if a disjunction is involved. This would be the simplest solution to implement and would behave correctly in all cases, at the cost of making the toplevel output a bit less pretty if there's a `when` with a disjunction.

**Comment by JanWielemaker** [comment]
Correctness is more important that pretty, that is easy :smile:    Correct and pretty is best, of course.  One way to get there is (I think) for a disjunction to know about the other branches and kill their attributes.   This would also fix call_residue_vars/2 for this case, no?   Seems you have more experience with this than I.   Any clue how much worse it would work in practice considering performance?   Next to loosing we might also gain by making the disjunctive attributes garbage when we commit, no?

**Comment by dgelessus** [comment]
> One way to get there is (I think) for a disjunction to know about the other branches and kill their attributes. This would also fix call_residue_vars/2 for this case, no?

The individual disjunction attributes already have information about all other parts of the disjunction (the second argument of `when:check_disj` is a list of all conditions in the disjunction), so this probably wouldn't be hard to implement. That would indeed also fix my original problem with the residual goals being shown even though the `when` has already fired. But it wouldn't affect the bug that currently happens when using `frozen/2`, because that bug only happens if the `when` *hasn't* fired yet.

I definitely want to try implementing this in the future and see how it affects performance and memory usage. But if possible I would like to get the bugs in the current implementatin fixed first, so that I can properly compare the implementations with/without deleting the unneeded attributes. I'll try implementing the fixes that I'm thinking of in the coming week.

**Comment by dgelessus** [comment]
> if you call `frozen/2` on a variable that appears in a `when` condition disjunction, the `when` effectively disappears and will not execute once the condition becomes true.

I've now run into a similar problem with CLPFD as well. I'll document this in more detail in #828, as this problem is apparently not specific to `when`. I'll also close this issue, because the original bug has been fixed in #832.
